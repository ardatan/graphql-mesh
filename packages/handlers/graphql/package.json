{
  "name": "@graphql-mesh/graphql",
  "version": "0.17.10",
  "sideEffects": false,
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "typescript": {
    "definition": "dist/index.d.ts"
  },
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "require": "./dist/*.js",
      "import": "./dist/*.mjs"
    }
  },
  "license": "MIT",
  "peerDependencies": {
    "graphql": "*"
  },
  "dependencies": {
    "@graphql-mesh/types": "0.42.0",
    "@graphql-mesh/utils": "0.12.0",
    "@graphql-tools/url-loader": "6.10.1",
    "@graphql-tools/wrap": "7.0.8"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
