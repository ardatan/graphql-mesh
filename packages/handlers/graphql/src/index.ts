import {
  GetMeshSourceOptions,
  MeshHandler,
  MeshSource,
  ResolverData,
  YamlConfig,
  KeyValueCache,
} from '@graphql-mesh/types';
import { UrlLoader } from '@graphql-tools/url-loader';
import {
  GraphQLSchema,
  buildClientSchema,
  introspectionFromSchema,
  buildSchema,
  IntrospectionQuery,
  parse,
  execute,
  GraphQLResolveInfo,
  DocumentNode,
  Kind,
  buildASTSchema,
} from 'graphql';
import { introspectSchema } from '@graphql-tools/wrap';
import {
  getInterpolatedHeadersFactory,
  ResolverDataBasedFactory,
  getHeadersObject,
  loadFromModuleExportExpression,
  getInterpolatedStringFactory,
  getCachedFetch,
  readFileOrUrlWithCache,
} from '@graphql-mesh/utils';
import { ExecutionParams, AsyncExecutor } from '@graphql-tools/delegate';

export interface GraphQLHandlerIntrospection {
  introspection?: IntrospectionQuery;
  sdl?: string;
}

const APOLLO_GET_SERVICE_DEFINITION_QUERY = /* GraphQL */ `
  query __ApolloGetServiceDefinition__ {
    _service {
      sdl
    }
  }
`;

export default class GraphQLHandler implements MeshHandler {
  private config: YamlConfig.GraphQLHandler;
  private baseDir: string;
  private cache: KeyValueCache<any>;
  private introspectionCache: GraphQLHandlerIntrospection;

  constructor({
    config,
    baseDir,
    cache,
    introspectionCache = {},
  }: GetMeshSourceOptions<YamlConfig.GraphQLHandler, GraphQLHandlerIntrospection>) {
    this.config = config;
    this.baseDir = baseDir;
    this.cache = cache;
    this.introspectionCache = introspectionCache;
  }

  async getMeshSource(): Promise<MeshSource> {
    const { endpoint, schemaHeaders: configHeaders, introspection } = this.config;
    const customFetch = getCachedFetch(this.cache);

    if (endpoint.endsWith('.js') || endpoint.endsWith('.ts')) {
      // Loaders logic should be here somehow
      const schemaOrStringOrDocumentNode = await loadFromModuleExportExpression<GraphQLSchema | string | DocumentNode>(
        endpoint,
        { cwd: this.baseDir }
      );
      let schema: GraphQLSchema;
      if (schemaOrStringOrDocumentNode instanceof GraphQLSchema) {
        schema = schemaOrStringOrDocumentNode;
      } else if (typeof schemaOrStringOrDocumentNode === 'string') {
        schema = buildSchema(schemaOrStringOrDocumentNode);
      } else if (
        typeof schemaOrStringOrDocumentNode === 'object' &&
        schemaOrStringOrDocumentNode?.kind === Kind.DOCUMENT
      ) {
        schema = buildASTSchema(schemaOrStringOrDocumentNode);
      } else {
        throw new Error(
          `Provided file '${endpoint} exports an unknown type: ${typeof schemaOrStringOrDocumentNode}': expected GraphQLSchema, SDL or DocumentNode.`
        );
      }
      return {
        schema,
      };
    } else if (endpoint.endsWith('.graphql')) {
      const rawSDL = await readFileOrUrlWithCache<string>(endpoint, this.cache, {
        cwd: this.baseDir,
        allowUnknownExtensions: true,
      });
      const schema = buildSchema(rawSDL);
      return {
        schema,
      };
    }
    const urlLoader = new UrlLoader();
    const getExecutorAndSubscriberForParams = (
      params: ExecutionParams,
      headersFactory: ResolverDataBasedFactory<Headers>,
      endpointFactory: ResolverDataBasedFactory<string>
    ) => {
      const resolverData: ResolverData = {
        root: {},
        args: params.variables,
        context: params.context,
      };
      const headers = getHeadersObject(headersFactory(resolverData));
      const endpoint = endpointFactory(resolverData);
      return urlLoader.getExecutorAndSubscriberAsync(endpoint, {
        customFetch,
        ...this.config,
        headers,
      });
    };
    let schemaHeaders =
      typeof configHeaders === 'string'
        ? await loadFromModuleExportExpression(configHeaders, { cwd: this.baseDir })
        : configHeaders;
    if (typeof schemaHeaders === 'function') {
      schemaHeaders = schemaHeaders();
    }
    if (schemaHeaders && 'then' in schemaHeaders) {
      schemaHeaders = await schemaHeaders;
    }
    const schemaHeadersFactory = getInterpolatedHeadersFactory(schemaHeaders || {});
    const introspectionExecutor: AsyncExecutor = async (params): Promise<any> => {
      const { executor } = await getExecutorAndSubscriberForParams(params, schemaHeadersFactory, () => endpoint);
      return executor(params);
    };
    if (!this.introspectionCache.introspection) {
      if (introspection) {
        const result = await urlLoader.handleSDLAsync(introspection, {
          customFetch,
          ...this.config,
          headers: schemaHeaders,
        });
        this.introspectionCache.introspection = introspectionFromSchema(result.schema);
      } else {
        this.introspectionCache.introspection = introspectionFromSchema(await introspectSchema(introspectionExecutor));
      }
    }
    const nonExecutableSchema = buildClientSchema(this.introspectionCache.introspection);
    const operationHeadersFactory = getInterpolatedHeadersFactory(this.config.operationHeaders);
    const endpointFactory = getInterpolatedStringFactory(endpoint);
    const queryType = nonExecutableSchema.getQueryType();
    const queryTypeFieldMap = queryType.getFields();
    if ('_service' in queryTypeFieldMap) {
      const _serviceField = queryTypeFieldMap._service;
      if ('resolve' in _serviceField) {
        _serviceField.resolve = async () => {
          if (!this.introspectionCache.sdl) {
            const sdlQueryResult = await introspectionExecutor({
              document: parse(APOLLO_GET_SERVICE_DEFINITION_QUERY),
            });
            this.introspectionCache.sdl = sdlQueryResult?.data?._service?.sdl;
          }
          return {
            sdl: this.introspectionCache.sdl,
          };
        };
      }
    }
    const isSdlQuery = (info: GraphQLResolveInfo) =>
      info.fieldName === '_service' &&
      info.fieldNodes[0].selectionSet.selections[0].kind === 'Field' &&
      info.fieldNodes[0].selectionSet.selections[0].name.value === 'sdl';
    return {
      schema: nonExecutableSchema,
      executor: async params => {
        if (isSdlQuery(params.info)) {
          return execute(nonExecutableSchema, params.document);
        }
        const { executor } = await getExecutorAndSubscriberForParams(params, operationHeadersFactory, endpointFactory);
        return executor(params) as any;
      },
      subscriber: async params => {
        const { subscriber } = await getExecutorAndSubscriberForParams(
          params,
          operationHeadersFactory,
          endpointFactory
        );
        return subscriber(params);
      },
      batch: 'batch' in this.config ? this.config.batch : true,
    };
  }
}
