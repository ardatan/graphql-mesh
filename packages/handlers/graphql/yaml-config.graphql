extend type Handler {
  """
  Handler for remote/local/third-party GraphQL schema
  """
  graphql: GraphQLHandler
}

type GraphQLHandler @md {
  """
  JSON object representing the Headers to add to the runtime of the API calls
  """
  headers: JSON
  """
  A url to your remote GraphQL endpoint
  """
  endpoint: String!
}
