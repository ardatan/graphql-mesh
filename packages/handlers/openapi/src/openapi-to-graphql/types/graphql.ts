// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
