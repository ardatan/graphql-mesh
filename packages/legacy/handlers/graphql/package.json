{
  "name": "@graphql-mesh/graphql",
  "version": "0.103.11",
  "type": "module",
  "repository": {
    "type": "git",
    "url": "ardatan/graphql-mesh",
    "directory": "packages/legacy/handlers/graphql"
  },
  "license": "MIT",
  "engines": {
    "node": ">=16.0.0"
  },
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "exports": {
    ".": {
      "require": {
        "types": "./dist/typings/index.d.cts",
        "default": "./dist/cjs/index.js"
      },
      "import": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      },
      "default": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      }
    },
    "./package.json": "./package.json"
  },
  "typings": "dist/typings/index.d.ts",
  "peerDependencies": {
    "graphql": "*"
  },
  "dependencies": {
    "@graphql-mesh/cross-helpers": "^0.4.9",
    "@graphql-mesh/store": "^0.103.10",
    "@graphql-mesh/string-interpolation": "^0.5.7",
    "@graphql-mesh/types": "^0.103.10",
    "@graphql-mesh/utils": "^0.103.10",
    "@graphql-tools/delegate": "^10.0.28",
    "@graphql-tools/federation": "^3.0.0",
    "@graphql-tools/merge": "^9.0.12",
    "@graphql-tools/url-loader": "^8.0.9",
    "@graphql-tools/utils": "^10.6.0",
    "lodash.get": "^4.4.2",
    "tslib": "^2.4.0"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "sideEffects": false,
  "typescript": {
    "definition": "dist/typings/index.d.ts"
  }
}
