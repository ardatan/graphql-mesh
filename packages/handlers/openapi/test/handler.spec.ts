import { EventEmitter } from 'tsee';
import { printSchema } from 'graphql';
import handler from '../src';
import { InMemoryLRUCache } from '@graphql-mesh/cache-inmemory-lru';
import { resolve } from 'path';

describe('openapi', () => {
  it('should create a GraphQL schema from a simple local swagger file', async () => {
    const source = await handler.getMeshSource({
      name: 'Instagram',
      config: {
        source: resolve(__dirname, './fixtures/instagram.json'),
      },
      hooks: new EventEmitter(),
      cache: new InMemoryLRUCache(),
    });

    expect(printSchema(source.schema)).toMatchSnapshot();
  });
});
