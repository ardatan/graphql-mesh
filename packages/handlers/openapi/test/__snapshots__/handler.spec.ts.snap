// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi should create a GraphQL schema from a simple local swagger file 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query {
  \\"\\"\\"
  Get recent media from a geography subscription that you created.
  **Note:** You can only access Geographies that were explicitly created by your OAuth client. Check the
  Geography Subscriptions section of the [real-time updates page](https://instagram.com/developer/realtime/).
  When you create a subscription to some geography that you define, you will be returned a unique \`geo-id\` that
  can be used in this query. To backfill photos from the location covered by this geography, use the
  [media search endpoint](https://instagram.com/developer/endpoints/media/).
  **Warning:** [Deprecated](http://instagram.com/developer/changelog/) for Apps created **on or after** Nov 17, 2015
  
  
  Equivalent to GET /geographies/{geo-id}/media/recent
  \\"\\"\\"
  getGeographiesGeoIdMediaRecent(
    \\"\\"\\"Max number of media to return.\\"\\"\\"
    count: Int
    \\"\\"\\"The geography ID.\\"\\"\\"
    geoId: String!
    \\"\\"\\"Return media before this \`min_id\`.\\"\\"\\"
    minId: String
  ): MediaListResponse
  \\"\\"\\"
  Get information about a location.
  
  Equivalent to GET /locations/{location-id}
  \\"\\"\\"
  getLocationsLocationId(
    \\"\\"\\"The location ID.\\"\\"\\"
    locationId: String!
  ): LocationInfoResponse
  \\"\\"\\"
  Get a list of recent media objects from a given location.
  
  Equivalent to GET /locations/{location-id}/media/recent
  \\"\\"\\"
  getLocationsLocationIdMediaRecent(
    \\"\\"\\"The location ID.\\"\\"\\"
    locationId: String!
    \\"\\"\\"Return media after this \`max_id\`.\\"\\"\\"
    maxId: String
    \\"\\"\\"Return media before this UNIX timestamp.\\"\\"\\"
    maxTimestamp: Float
    \\"\\"\\"Return media before this \`min_id\`.\\"\\"\\"
    minId: String
    \\"\\"\\"Return media after this UNIX timestamp.\\"\\"\\"
    minTimestamp: Float
  ): MediaListResponse
  \\"\\"\\"
  Search for a location by geographic coordinate.
  
  Equivalent to GET /locations/search
  \\"\\"\\"
  getLocationsSearch(
    \\"\\"\\"Default is 1000m (distance=1000), max distance is 5000.\\"\\"\\"
    distance: Int
    \\"\\"\\"Returns a location mapped off of a Facebook places id. If used, a Foursquare id and \`lat\`, \`lng\` are not required.\\"\\"\\"
    facebookPlacesId: String
    \\"\\"\\"
    Returns a location mapped off of a foursquare v1 api location id. If used, you are not required to use
    \`lat\` and \`lng\`. Note that this method is deprecated; you should use the new foursquare IDs with V2 of their API.
    
    \\"\\"\\"
    foursquareId: String
    \\"\\"\\"
    Returns a location mapped off of a foursquare v2 api location id. If used, you are not required to use
    \`lat\` and \`lng\`.
    
    \\"\\"\\"
    foursquareV2Id: String
    \\"\\"\\"Latitude of the center search coordinate. If used, \`lng\` is required.\\"\\"\\"
    lat: Float
    \\"\\"\\"Longitude of the center search coordinate. If used, \`lat\` is required.\\"\\"\\"
    lng: Float
  ): LocationSearchResponse
  \\"\\"\\"
  Get information about a media object. The returned type key will allow you to differentiate between image and
  video media.
  **Note:** if you authenticate with an OAuth Token, you will receive the user_has_liked key which quickly tells
  you whether the current user has liked this media item.
  
  
  Equivalent to GET /media/{media-id}
  \\"\\"\\"
  getMediaMediaId(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): MediaEntryResponse
  \\"\\"\\"
  Get a list of recent comments on a media object.
  
  Equivalent to GET /media/{media-id}/comments
  \\"\\"\\"
  getMediaMediaIdComments(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): CommentsResponse
  \\"\\"\\"
  Get a list of users who have liked this media.
  
  Equivalent to GET /media/{media-id}/likes
  \\"\\"\\"
  getMediaMediaIdLikes(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): UsersInfoResponse
  \\"\\"\\"
  Get a list of what media is most popular at the moment. Can return mix of \`image\` and \`video\` types.
  **Warning:** [Deprecated](http://instagram.com/developer/changelog/) for Apps created **on or after** Nov 17, 2015
  
  
  Equivalent to GET /media/popular
  \\"\\"\\"
  getMediaPopular: MediaSearchResponse
  \\"\\"\\"
  Search for media in a given area. The default time span is set to 5 days. The time span must not exceed 7 days.
  Defaults time stamps cover the last 5 days. Can return mix of \`image\` and \`video\` types.
  
  
  Equivalent to GET /media/search
  \\"\\"\\"
  getMediaSearch(
    \\"\\"\\"Default is 1km (distance=1000), max distance is 5km.\\"\\"\\"
    distance: Int
    \\"\\"\\"Latitude of the center search coordinate. If used, \`lng\` is required.\\"\\"\\"
    lat: Float!
    \\"\\"\\"Longitude of the center search coordinate. If used, \`lat\` is required.\\"\\"\\"
    lng: Float!
    \\"\\"\\"A unix timestamp. All media returned will be taken earlier than this timestamp.\\"\\"\\"
    maxTimestamp: Float
    \\"\\"\\"A unix timestamp. All media returned will be taken later than this timestamp.\\"\\"\\"
    minTimestamp: Float
  ): MediaSearchResponse
  \\"\\"\\"
  This endpoint returns the same response as \`GET /media/{media-id}\`.
  A media object's shortcode can be found in its shortlink URL. An example shortlink is
  \`http://instagram.com/p/D/\`, its corresponding shortcode is \`D\`.
  
  
  Equivalent to GET /media/shortcode/{shortcode}
  \\"\\"\\"
  getMediaShortcodeShortcode(
    \\"\\"\\"The short code of the media resource.\\"\\"\\"
    shortcode: String!
  ): MediaEntryResponse
  \\"\\"\\"
  Search for tags by name.
  
  Equivalent to GET /tags/search
  \\"\\"\\"
  getTagsSearch(
    \\"\\"\\"A valid tag name without a leading \\\\#. (eg. snowy, nofilter)\\"\\"\\"
    q: String!
  ): TagSearchResponse
  \\"\\"\\"
  Get information about a tag object.
  
  Equivalent to GET /tags/{tag-name}
  \\"\\"\\"
  getTagsTagName(
    \\"\\"\\"The tag name.\\"\\"\\"
    tagName: String!
  ): TagInfoResponse
  \\"\\"\\"
  Get a list of recently tagged media. Use the \`max_tag_id\` and \`min_tag_id\` parameters in the pagination
  response to paginate through these objects.
  
  
  Equivalent to GET /tags/{tag-name}/media/recent
  \\"\\"\\"
  getTagsTagNameMediaRecent(
    \\"\\"\\"Count of tagged media to return.\\"\\"\\"
    count: Int
    \\"\\"\\"Return media after this \`max_tag_id\`.\\"\\"\\"
    maxTagId: String
    \\"\\"\\"Return media before this \`min_tag_id\`.\\"\\"\\"
    minTagId: String
    \\"\\"\\"The tag name.\\"\\"\\"
    tagName: String!
  ): TagMediaListResponse
  \\"\\"\\"
  Search for a user by name.
  
  Equivalent to GET /users/search
  \\"\\"\\"
  getUsersSearch(
    \\"\\"\\"Number of users to return.\\"\\"\\"
    count: Int
    \\"\\"\\"A query string.\\"\\"\\"
    q: String!
  ): UsersInfoResponse
  \\"\\"\\"
  See the authenticated user's feed.
  **Warning:** [Deprecated](http://instagram.com/developer/changelog/) for Apps created **on or after** Nov 17, 2015
  
  
  Equivalent to GET /users/self/feed
  \\"\\"\\"
  getUsersSelfFeed(
    \\"\\"\\"Count of media to return.\\"\\"\\"
    count: Int
    \\"\\"\\"Return media earlier than this \`max_id\`.\\"\\"\\"
    maxId: String
    \\"\\"\\"Return media later than this \`min_id\`.\\"\\"\\"
    minId: String
  ): MediaListResponse
  \\"\\"\\"
  See the list of media liked by the authenticated user. Private media is returned as long as the authenticated
  user has permission to view that media. Liked media lists are only available for the currently authenticated
  user.
  
  
  Equivalent to GET /users/self/media/liked
  \\"\\"\\"
  getUsersSelfMediaLiked(
    \\"\\"\\"Count of media to return.\\"\\"\\"
    count: Int
    \\"\\"\\"Return media liked before this id.\\"\\"\\"
    maxLikeId: String
  ): MediaListResponse
  \\"\\"\\"
  List the users who have requested this user's permission to follow.
  
  Equivalent to GET /users/self/requested-by
  \\"\\"\\"
  getUsersSelfRequestedBy: UsersInfoResponse
  \\"\\"\\"
  Get basic information about a user. To get information about the owner of the access token, you can use
  **self** instead of the \`user-id\`.
  Security scope \`public_content\` is required to read information about other users.
  
  
  Equivalent to GET /users/{user-id}
  \\"\\"\\"
  getUsersUserId(
    \\"\\"\\"The ID of a user to get information about, or **self** to retrieve information about authenticated user.\\"\\"\\"
    userId: String!
  ): UserResponse
  \\"\\"\\"
  Get the list of users this user is followed by. To get users followed by the owner of the access token, you
  can use **self** instead of the \`user-id\`.
  
  
  Equivalent to GET /users/{user-id}/followed-by
  \\"\\"\\"
  getUsersUserIdFollowedBy(
    \\"\\"\\"The ID of a user, or **self** to retrieve information about authenticated user.\\"\\"\\"
    userId: String!
  ): UsersPagingResponse
  \\"\\"\\"
  Get the list of users this user follows. To get follows of the owner of the access token, you can use **self**
  instead of the \`user-id\`.
  
  
  Equivalent to GET /users/{user-id}/follows
  \\"\\"\\"
  getUsersUserIdFollows(
    \\"\\"\\"The ID of a user, or **self** to retrieve information about authenticated user.\\"\\"\\"
    userId: String!
  ): UsersPagingResponse
  \\"\\"\\"
  Get the most recent media published by a user. To get the most recent media published by the owner of the
  access token, you can use **self** instead of the \`user-id\`.
  Security scope \`public_content\` is required to read information about other users.
  
  
  Equivalent to GET /users/{user-id}/media/recent
  \\"\\"\\"
  getUsersUserIdMediaRecent(
    \\"\\"\\"Count of media to return.\\"\\"\\"
    count: Int
    \\"\\"\\"Return media earlier than this \`max_id\`.\\"\\"\\"
    maxId: String
    \\"\\"\\"Return media before this UNIX timestamp.\\"\\"\\"
    maxTimestamp: Float
    \\"\\"\\"Return media later than this \`min_id\`.\\"\\"\\"
    minId: String
    \\"\\"\\"Return media after this UNIX timestamp.\\"\\"\\"
    minTimestamp: Float
    \\"\\"\\"The ID of a user to get recent media of, or **self** to retrieve media of authenticated user.\\"\\"\\"
    userId: String!
  ): MediaListResponse
  \\"\\"\\"
  Get information about a relationship to another user.
  
  Equivalent to GET /users/{user-id}/relationship
  \\"\\"\\"
  getUsersUserIdRelationship(
    \\"\\"\\"The ID of a user to get information about.\\"\\"\\"
    userId: String!
  ): RelationshipResponse
}

type MediaListResponse {
  \\"\\"\\"List of media entries\\"\\"\\"
  data: [MediaEntry]
  meta: MetaData
  pagination: IdPaginationInfo
}

type MediaEntry {
  \\"\\"\\"??? Unknown ???\\"\\"\\"
  attribution: String
  caption: CaptionData
  comments: CommentsCollection
  \\"\\"\\"Media creation UNIX timestamp\\"\\"\\"
  createdTime: String
  \\"\\"\\"Filter of this media entry\\"\\"\\"
  filter: String
  \\"\\"\\"ID of a media entry\\"\\"\\"
  id: String
  images: ImagesData
  likes: LikesCollection
  \\"\\"\\"Fixed URL of this media entry\\"\\"\\"
  link: String
  location: LocationInfo
  \\"\\"\\"List of tags assigned to this media\\"\\"\\"
  tags: [String]
  \\"\\"\\"Type of this media entry\\"\\"\\"
  type: Type
  user: UserShortInfo
  \\"\\"\\"Indicates whether authenticated user has liked this media or not\\"\\"\\"
  userHasLiked: Boolean
  \\"\\"\\"Users located on this media entry\\"\\"\\"
  usersInPhoto: [UserInPhoto]
  videos: VideosData
}

type CaptionData {
  \\"\\"\\"Caption creation UNIX timestamp\\"\\"\\"
  createdTime: String
  from: UserShortInfo
  \\"\\"\\"ID of this caption\\"\\"\\"
  id: String
  \\"\\"\\"Caption text\\"\\"\\"
  text: String
}

type UserShortInfo {
  \\"\\"\\"User full name\\"\\"\\"
  fullName: String
  \\"\\"\\"User ID\\"\\"\\"
  id: String
  \\"\\"\\"URL to user profile picture\\"\\"\\"
  profilePicture: String
  \\"\\"\\"User name, nickname\\"\\"\\"
  username: String
}

type CommentsCollection {
  \\"\\"\\"Nember of comments available, data does not necessary contain all comments\\"\\"\\"
  count: Int
  \\"\\"\\"Collection of comment entries; **warning:** deprecated for Apps [created on or after Nov 17, 2015](http://instagram.com/developer/changelog/)\\"\\"\\"
  data: [CommentEntry]
}

type CommentEntry {
  \\"\\"\\"Comment creation UNIX timestamp\\"\\"\\"
  createdTime: String
  from: UserShortInfo
  \\"\\"\\"ID of this comment\\"\\"\\"
  id: String
  \\"\\"\\"Text of the comment\\"\\"\\"
  text: String
}

type ImagesData {
  lowResolution: ImageInfo
  standardResolution: ImageInfo
  thumbnail: ImageInfo
}

type ImageInfo {
  \\"\\"\\"Image/video height in pixels\\"\\"\\"
  height: Int
  \\"\\"\\"URL of the image/video resource\\"\\"\\"
  url: String
  \\"\\"\\"Image/video width in pixels\\"\\"\\"
  width: Int
}

type LikesCollection {
  \\"\\"\\"Nember of likes available, data does not necessary contain all comments\\"\\"\\"
  count: Int
  \\"\\"\\"Collection of users who liked; **warning:** deprecated for Apps [created on or after Nov 17, 2015](http://instagram.com/developer/changelog/)\\"\\"\\"
  data: [UserShortInfo]
}

type LocationInfo {
  \\"\\"\\"ID of this location (in some responses it has a type of 'integer')\\"\\"\\"
  id: String
  \\"\\"\\"Location latitude\\"\\"\\"
  latitude: Float
  \\"\\"\\"Location longitude\\"\\"\\"
  longitude: Float
  \\"\\"\\"Location name\\"\\"\\"
  name: String
}

enum Type {
  IMAGE
  VIDEO
}

type UserInPhoto {
  position: Position
  user: UserShortInfo
}

type Position {
  \\"\\"\\"X position (horizontal)\\"\\"\\"
  x: Float
  \\"\\"\\"Y position (vertical)\\"\\"\\"
  y: Float
}

type VideosData {
  lowResolution: ImageInfo
  standardResolution: ImageInfo
}

type MetaData {
  \\"\\"\\"HTTP result code\\"\\"\\"
  code: Int
}

type IdPaginationInfo {
  \\"\\"\\"The max ID of the next page\\"\\"\\"
  nextMaxId: String
  \\"\\"\\"URL to retrieve next page of entries\\"\\"\\"
  nextUrl: String
}

type LocationInfoResponse {
  data: LocationInfo
  meta: MetaData
}

type LocationSearchResponse {
  \\"\\"\\"List of found locations\\"\\"\\"
  data: [LocationInfo]
  meta: MetaData
}

type MediaEntryResponse {
  data: MediaEntry
  meta: MetaData
}

type CommentsResponse {
  \\"\\"\\"Collection of comments\\"\\"\\"
  data: [CommentEntry]
  meta: MetaData
}

type UsersInfoResponse {
  \\"\\"\\"User short information entries\\"\\"\\"
  data: [UserShortInfo]
  meta: MetaData
}

type MediaSearchResponse {
  \\"\\"\\"Found media entries; some end-points do not return likes informtaion\\"\\"\\"
  data: [MediaEntry]
  meta: MetaData
}

type TagSearchResponse {
  \\"\\"\\"List of found tags with brief statistics\\"\\"\\"
  data: [TagInfo]
  meta: MetaData
}

type TagInfo {
  \\"\\"\\"Overall number of media entries taged with this name\\"\\"\\"
  mediaCount: Float
  \\"\\"\\"Tag name\\"\\"\\"
  name: String
}

type TagInfoResponse {
  data: TagInfo
  meta: MetaData
}

type TagMediaListResponse {
  \\"\\"\\"List of media entries with this tag\\"\\"\\"
  data: [MediaEntry]
  meta: MetaData
  pagination: TagPaginationInfo
}

type TagPaginationInfo {
  \\"\\"\\"The deprication warning, if information is available\\"\\"\\"
  deprecationWarning: String
  \\"\\"\\"The min ID of a tag for the next page\\"\\"\\"
  minTagId: String
  \\"\\"\\"Depricated. Use min_tag_id instead\\"\\"\\"
  nextMaxId: String
  \\"\\"\\"The max ID of a tag for the next page\\"\\"\\"
  nextMaxTagId: String
  \\"\\"\\"Depricated. Use max_tag_id instead\\"\\"\\"
  nextMinId: String
  \\"\\"\\"URL to retrieve next page of entries\\"\\"\\"
  nextUrl: String
}

type UserResponse {
  data: UserInfo
  meta: MetaData
}

type UserInfo {
  \\"\\"\\"User biography\\"\\"\\"
  bio: String
  counts: UserCounts
  \\"\\"\\"User full name\\"\\"\\"
  fullName: String
  \\"\\"\\"User ID\\"\\"\\"
  id: String
  \\"\\"\\"URL to user profile picture\\"\\"\\"
  profilePicture: String
  \\"\\"\\"User name, nickname\\"\\"\\"
  username: String
  \\"\\"\\"URL to user web-site\\"\\"\\"
  website: String
}

type UserCounts {
  \\"\\"\\"Number of followers of this user\\"\\"\\"
  followedBy: Int
  \\"\\"\\"Number of users followed by this user\\"\\"\\"
  follows: Int
  \\"\\"\\"Number of user media\\"\\"\\"
  media: Int
}

type UsersPagingResponse {
  \\"\\"\\"List of user short information entries\\"\\"\\"
  data: [UserShortInfo]
  meta: MetaData
  pagination: CursorPaginationInfo
}

type CursorPaginationInfo {
  \\"\\"\\"The cursor ID of the next page\\"\\"\\"
  nextCursor: String
  \\"\\"\\"URL to retrieve next page of entries\\"\\"\\"
  nextUrl: String
}

type RelationshipResponse {
  data: RelationshipInfo
  meta: MetaData
}

type RelationshipInfo {
  \\"\\"\\"Status of incoming relationship\\"\\"\\"
  incomingStatus: IncomingStatus
  \\"\\"\\"Status of outgoing relationship\\"\\"\\"
  outgoingStatus: OutgoingStatus
  \\"\\"\\"Indicates whether target user is private or not\\"\\"\\"
  targetUserIsPrivate: Boolean
}

enum IncomingStatus {
  NONE
  FOLLOWED_BY
  REQUESTED_BY
}

enum OutgoingStatus {
  NONE
  FOLLOWS
  REQUESTED
}

type Mutation {
  \\"\\"\\"
  Remove a comment either on the authenticated user's media object or authored by the authenticated user.
  
  
  Equivalent to DELETE /media/{media-id}/comments/{comment-id}
  \\"\\"\\"
  deleteMediaMediaIdCommentsCommentId(
    \\"\\"\\"The ID of the comment entry.\\"\\"\\"
    commentId: String!
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): StatusResponse
  \\"\\"\\"
  Remove a like on this media by the currently authenticated user.
  
  Equivalent to DELETE /media/{media-id}/likes
  \\"\\"\\"
  deleteMediaMediaIdLikes(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): StatusResponse
  \\"\\"\\"
  Create a comment on a media object with the following rules:
  
    * The total length of the comment cannot exceed 300 characters.
    * The comment cannot contain more than 4 hashtags.
    * The comment cannot contain more than 1 URL.
    * The comment cannot consist of all capital letters.
  
  
  Equivalent to POST /media/{media-id}/comments
  \\"\\"\\"
  postMediaMediaIdComments(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
    \\"\\"\\"Text to post as a comment on the media object as specified in \`media-id\`.\\"\\"\\"
    text: String!
  ): StatusResponse
  \\"\\"\\"
  Set a like on this media by the currently authenticated user.
  
  Equivalent to POST /media/{media-id}/likes
  \\"\\"\\"
  postMediaMediaIdLikes(
    \\"\\"\\"The ID of the media resource.\\"\\"\\"
    mediaId: String!
  ): StatusResponse
  \\"\\"\\"
  Modify the relationship between the current user and the target user.
  
  Equivalent to POST /users/{user-id}/relationship
  \\"\\"\\"
  postUsersUserIdRelationship(
    \\"\\"\\"Type of action to apply for relationship with the user.\\"\\"\\"
    action: Action!
    \\"\\"\\"The ID of the target user.\\"\\"\\"
    userId: String!
  ): RelationshipPostResponse
}

type StatusResponse {
  \\"\\"\\"No data - 'null'\\"\\"\\"
  data: String
  meta: MetaData
}

type RelationshipPostResponse {
  data: RelationshipStatus
  meta: MetaData
}

type RelationshipStatus {
  \\"\\"\\"Status of outgoing relationship\\"\\"\\"
  outgoingStatus: OutgoingStatus
}

enum Action {
  FOLLOW
  UNFOLLOW
  BLOCK
  UNBLOCK
  APPROVE
  IGNORE
}
"
`;
